README.md
# Varnishslog &emsp; [![Latest Version](https://img.shields.io/crates/v/varnishslog.svg)](https://crates.io/crates/varnishslog)

**`varnishslog` reads Varnish Cache VSL (binary) log stream and produces structured log records in JSON format.**

---

`varnishslog` reads Varnish VSL (binary) log stream and builds a model of access session:

![Session Record Model](session_record_model.png)

This data is then flattened down and serialised as JSON like this one:

```json
{
  "record_type": "client_request",
  "vxid": 32802,
  "remote_address": {
    "ip": "127.0.0.1",
    "port": 38650
  },
  "session_timestamp": 1476361595.724506,
  "start_timestamp": 1476361595.724543,
  "end_timestamp": 1476361595.869836,
  "handling": "miss",
  "request": {
    "protocol": "HTTP/1.1",
    "method": "GET",
    "url": "/",
    "headers": {
      "User-Agent": [
        "curl/7.40.0"
      ],
      "Accept": [
        "*/*"
      ],
      "Host": [
        "foo.net"
      ],
      "X-Forwarded-For": [
        "127.0.0.1"
      ]
    }
  },
  "response": {
    "status": 200,
    "reason": "OK",
    "protocol": "HTTP/1.1",
    "headers": {
      "Content-Type": [
        "text/html; charset=utf-8"
      ],
      "Date": [
        "Thu, 13 Oct 2016 12:26:35 GMT"
      ],
      "Etag": [
        "W/\"57973331\""
      ],
      "Vary": [
        "Accept-Encoding, Cookie, User-Agent"
      ],
      "Cache-Control": [
        "no-cache"
      ],
      "Accept-Ranges": [
        "bytes"
      ],
      "Transfer-Encoding": [
        "chunked"
      ],
      "Connection": [
        "keep-alive"
      ]
    }
  },
  "backend_access": {
    "vxid": 32803,
    "start_timestamp": 1476361595.724695,
    "end_timestamp": 1476361595.869323,
    "handling": "fetch",
    "request": {
      "protocol": "HTTP/1.1",
      "method": "GET",
      "url": "/",
      "headers": {
        "User-Agent": [
          "curl/7.40.0"
        ],
        "Accept": [
          "*/*"
        ],
        "Host": [
          "foo.net"
        ],
        "Accept-Encoding": [
          "gzip"
        ]
      }
    },
    "response": {
      "status": 200,
      "reason": "OK",
      "protocol": "HTTP/1.1",
      "headers": {
        "Cache-Control": [
          "private, must-revalidate, s-maxage=3644"
        ],
        "Content-Type": [
          "text/html; charset=utf-8"
        ],
        "Vary": [
          "X-Accept-Encoding"
        ],
        "Content-Encoding": [
          "gzip"
        ],
        "Etag": [
          "\"57973331\""
        ],
        "Content-Length": [
          "183806"
        ],
        "Accept-Ranges": [
          "bytes"
        ],
        "Date": [
          "Thu, 13 Oct 2016 12:26:35 GMT"
        ],
        "Age": [
          "0"
        ],
        "Connection": [
          "keep-alive"
        ],
        "Via": [
          "1.1 varnish (Varnish)"
        ]
      }
    },
    "send_duration": 0.000744,
    "wait_duration": 0.138199,
    "ttfb_duration": 0.138943,
    "fetch_duration": 0.144628,
    "sent_header_bytes": 539,
    "sent_body_bytes": 0,
    "sent_total_bytes": 539,
    "recv_header_bytes": 509,
    "recv_body_bytes": 183806,
    "recv_total_bytes": 184315,
    "retry": 0,
    "backend_connection": {
      "fd": 30,
      "name": "boot.origin_primary",
      "remote_address": {
        "ip": "10.1.1.333",
        "port": 80
      },
      "local_address": {
        "ip": "10.1.1.222",
        "port": 53964
      }
    },
    "cache_object": {
      "storage_type": "file",
      "storage_name": "s0",
      "ttl_duration": 3644.0,
      "grace_duration": 259200.0,
      "keep_duration": 0.0,
      "since_timestamp": 1476361596.0,
      "origin_timestamp": 1476361596.0,
      "fetch_mode": "length",
      "fetch_streamed": true,
      "response": {
        "status": 200,
        "reason": "OK",
        "protocol": "HTTP/1.1",
        "headers": {
          "Cache-Control": [
            "private, must-revalidate, s-maxage=3644"
          ],
          "Content-Type": [
            "text/html; charset=utf-8"
          ],
          "Vary": [
            "Accept-Encoding"
          ],
          "Content-Encoding": [
            "gzip"
          ],
          "Etag": [
            "\"57973331\""
          ],
          "Content-Length": [
            "183806"
          ],
          "Date": [
            "Thu, 13 Oct 2016 12:26:35 GMT"
          ],
          "Via": [
            "1.1 varnish (Varnish)"
          ]
        }
      }
    },
    "compression": {
      "operation": "Gunzip-test",
      "bytes_in": 183806,
      "bytes_out": 489611
    },
    "log": {
      "vars": {
        "decision": "Cacheable",
        "esi_parsed": "false",
        "compressable": "false",
        "cache_class": "Cache-Control"
      },
      "messages": [],
      "acl_matched": [],
      "acl_not_matched": []
    }
  },
  "process_duration": 0.0,
  "fetch_duration": 0.139311,
  "ttfb_duration": 0.139434,
  "serve_duration": 0.145293,
  "recv_header_bytes": 106,
  "recv_body_bytes": 0,
  "recv_total_bytes": 106,
  "sent_header_bytes": 283,
  "sent_body_bytes": 489611,
  "sent_total_bytes": 489894,
  "esi_count": 0,
  "compression": {
    "operation": "Gunzip",
    "bytes_in": 183806,
    "bytes_out": 489611
  },
  "restart_count": 0,
  "log": {
    "vars": {
      "client_device": "D",
      "client_country": "Unknown",
      "request_id": "rid-32802",
    },
    "messages": [
      "geoip2.lookup: No entry for this IP address (127.0.0.1)",
      "RES_MODE 48"
    ],
    "acl_matched": [
      "trusted_networks",
      "unthrottled_clients"
    ],
    "acl_not_matched": [
      "forbidden_clients"
    ]
  }
}
```

It can also prepend it with NCSA like log line to make it more compatible with standard access log processing tools.

Compatibility
----

`varnishslog` will NOT work with Varnish v3 as there are significant differences in the VSL comparing to v4. I have no plans to support v3 as it is now discontinued.

`varnishslog` v0.4.x and earlier was tested with Varnish v4.1.3.

`varnishslog` v0.5.x was tested with Varnish v5.2.1.

`varnishslog` v0.6.x was tested with Varnish v6.1.1. Note that Varnish v6.1.0 has a VSL compatibility bug and `varnishslog` will not work correctly with this particular version of Varnish. Also note that logging of cache object response is disabled by default in Varnish going forward. To bring back "cached_object"."response" data add `-p vsl_mask=+ObjProtocol,+ObjStatus,+ObjReason,+ObjHeader` to `varnishd` parameters.

`varnishslog` v0.7.x was tested with Varnish v6.1.1 with HTTP/2 and PROXY v2 (hitch) protocol.

Usage
-----

To install it you will need recent Rust compiler (tested on rustc 1.27.2):
`cargo install varnishslog`

To run build from source with:
`cargo run --release`

Running:
`varnishlog -g raw -w /dev/stdout | varnishslog > access.log`

There are few output options and control over indexing of headers and log messages:

```
Varnish Structured Logger 0.4.0
Jakub Pastuszek <jpastuszek@gmail.com>
Reads Varnish VSL (binary) log stream and emits JSON log entries to standard output

USAGE:
    varnishslog [FLAGS] [OPTIONS] [<VSL_FILE>]

FLAGS:
    -h, --help                  Prints help information
    -I, --keep-raw-headers      Keep raw header name/value pairs; any indices are moved to top level
    -L, --keep-raw-log          Include raw log messages
    -i, --no-header-indexing    Do not make indices of request and response headers with normalized header names
    -l, --no-log-processing     Do not process VSL log into vars, messages and ACL matches
    -V, --version               Prints version information

OPTIONS:
    -d, --log-sepc <LOG_LEVEL_SPEC>                    Logging level specification, e.g: info [default: info]
    -o, --output-format <output>                       Format of the output [default: NcsaJson]  [values: Log, LogDebug,
                                                       RecordDebug, SessionDebug, Json, JsonPretty, NcsaJson]
    -s, --stat-epoch-interval <stat-epoch-interval>
            Log store stats every epoch interval (~seconds) on record insert [default: 600]
        --evict-factor <evict-factor>
            Remove that many records at a time when doing expiration or nucking (ratio to max-record-slots) [default: 0.01]
        --max-epoch-diff <max-epoch-diff>
            Consider record being built or correlated as stale after that many epoch (~seconds) elapsed [default: 14410]
        --max-record-slots <max-record-slots>
            Maximum number of records being built or correlated at the same time [default: 4000]
        --stream-buffer-size <stream-buffer-size>
            Size of stream buffer in bytes - must be bigger than biggest VSL record [default: 262144]

ARGS:
    <VSL_FILE>    VSL file to process (read from standard input if not specified)
```

Bench results
-------------

Run against v0.4.1 using 13480 VSL records making up 211 access records and 86 sessions and 86 client access records so about 157 VSL records per output client access record (benches/varnish20160816-4093-1xh1jbx808a493d5e74216e5.vsl).
This was run on MacBook Air (13-inch, Mid 2013) and build with rustc 1.14.0.

```
test log_session_record_ncsa_json ... bench:  14,940,553 ns/iter (+/- 4,323,366)
test log_session_record_json      ... bench:  15,029,912 ns/iter (+/- 3,922,717)
test log_session_record_json_raw  ... bench:   7,898,497 ns/iter (+/- 1,309,983)
test session_state                ... bench:   4,832,503 ns/iter (+/- 1,322,579)
test default_buffer_from_file     ... bench:     323,345 ns/iter (+/- 366,780)
```

* 5.8 K/s JSON records (indexed) or 902 K/s VSL records processed into serialized output
* 10.9 K/s JSON records (raw) or 1668 K/s VSL records processed into serialized output
* 17.8 K/s session records (correlated access records) or 2789 K/s VSL records processed (without serialization)
* 41.6 M/s VSL records extracted from binary stream (message not parsed)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [7.0.0] - 2018-11-23
### Fixed
- Added workaround for spurious End tag before SessionClose
- Properly handling case when Varnish is changing PIPE handling into PASS handling as PIPE is not supported with HTTP/2 to HTTP/1.1 translation
### Added
- Support for PROXY protocol client and server address logging (Proxy VSL tag)
- Remote address information is now taken from session for root client access records
- Client access records now have session information when available
- Support for bad request logging when there is no request information available
- SessionRecord now has close_reason string
- Logging of VCL_Error, ProxyGarbage and HttpGarbage messages

src/main.rs
extern crate flexi_logger;
extern crate time;
#[macro_use]
extern crate log;
#[macro_use]
extern crate quick_error;
#[macro_use]
extern crate clap;
extern crate varnishslog;

use std::io::{self, stdin, Read, Write};
use std::fs::File;
use std::error::Error;
use std::sync::atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT};
use std::num::Wrapping;
use std::time::Duration;
use std::thread;

use clap::{Arg, App};

use varnishslog::stream_buf::{StreamBuf, ReadStreamBuf, FillError, FillApplyError};
use varnishslog::vsl::record::VslRecord;
use varnishslog::vsl::record::parser::{binary_vsl_tag, vsl_record_v4};
use varnishslog::store::Config as StoreConfig;
use varnishslog::access_log::session_state::SessionState;
use varnishslog::access_log::record_state::RecordState;
use varnishslog::serialization::{log_client_record, Config, Format, OutputError, JsonError};

mod program;

static EPOCH: AtomicUsize = ATOMIC_USIZE_INIT;

fn spawn_epoch_timer() -> thread::JoinHandle<()> {
    thread::Builder::new()
        .name("epoch timer".to_string())
        .spawn(|| {
        loop {
            thread::sleep(Duration::from_secs(1));
            let epoch = Wrapping(EPOCH.load(Ordering::Relaxed) as u64) + Wrapping(1);
            EPOCH.store(epoch.0 as usize, Ordering::Relaxed);
        }
    }).expect("timer thread")
}

fn global_epoch(_epoch: Wrapping<u64>) -> Wrapping<u64> {
    Wrapping(EPOCH.load(Ordering::Relaxed) as u64)
}

quick_error! {
    #[derive(Debug)]
    pub enum ProcessingError {
        IO(err: io::Error) {
            display("IO Error while processing records: {}", err)
            description("I/O error")
            from()
        }
        InputBuffer(err: FillError) {
            display("Failed to fill parsing buffer: {}", err)
            description("Input buffer error")
        }
        Parsing(err: String) {
            display("Failed to parse VSL record: {}", err)
            description("Parser error")
        }
        Serialization(err: String) {
            display("Failed to serialize log record: {}", err)
            description("Serialization error")
        }
    }
}

impl<'b> From<FillApplyError<&'b[u8], u32>> for ProcessingError {
    fn from(err: FillApplyError<&'b[u8], u32>) -> ProcessingError {
        match err {
            FillApplyError::FillError(FillError::Io(err)) => ProcessingError::IO(err),
            FillApplyError::FillError(err) => ProcessingError::InputBuffer(err),
            // we need to convert it to string due to input reference
            FillApplyError::Parser(err) => ProcessingError::Parsing(format!("nom parser error: {}", err)),
        }
    }
}

impl From<OutputError> for ProcessingError {
    fn from(err: OutputError) -> ProcessingError {
        match err {
            OutputError::JsonSerialization(JsonError::Io(err)) |
            OutputError::Io(err) => ProcessingError::IO(err),
            err => ProcessingError::Serialization(format!("Serialization error: {}", err)),
        }
    }
}

impl ProcessingError {
    fn to_exit_code(&self) -> i32 {
        match *self {
            ProcessingError::IO(_) => 10,
            ProcessingError::InputBuffer(_) => 11,
            ProcessingError::Parsing(_) => 20,
            ProcessingError::Serialization(_) => 30,
        }
    }
}

impl ProcessingError {
    fn is_brokend_pipe(&self) -> bool {
        match *self {
            ProcessingError::IO(ref err) =>
                err.kind() == io::ErrorKind::UnexpectedEof || err.kind() == io::ErrorKind::BrokenPipe,
            _ => false
        }
    }
}

fn try_read_vsl_tag<R: Read>(stream: &mut ReadStreamBuf<R>) -> Result<(), ProcessingError> {
    loop {
        match try!(stream.fill_apply(binary_vsl_tag)) {
            None => continue,
            Some(Some(_)) => {
                info!("Found VSL tag");
                break
            }
            Some(_) => break,
        }
    }
    Ok(())
}

trait WriteRecord {
    fn write_record<W>(&mut self, record: VslRecord, output: &mut W) -> Result<(), ProcessingError> where W: Write;
    fn log_reports(&self) {}
}

fn process_vsl_records<R, W, P>(stream: &mut ReadStreamBuf<R>, mut writer: P, output: &mut W) -> Result<(), ProcessingError> where R: Read, W: Write, P: WriteRecord {
    loop {
        match stream.fill_apply(vsl_record_v4).map_err(ProcessingError::from) {
            Ok(None) => continue,
            Ok(Some(record)) => try!(writer.write_record(record, output)),
            Err(err) => {
                //TODO: need better tracking of orphan records and other stats
                if err.is_brokend_pipe() {
                    writer.log_reports();
                }
                return Err(err)
            }
        }
    }
}

fn process_vsl_stream<R, W>(input: R, mut output: W, stream_buf_size: usize, output_format: OutputFormat, config: Config, store_config: &StoreConfig) -> Result<(), ProcessingError> where R: Read, W: Write {
    //TODO: make buffer size configurable
    let mut stream = ReadStreamBuf::with_capacity(input, stream_buf_size);

    try!(try_read_vsl_tag(&mut stream));

    match output_format {
        OutputFormat::Log => process_vsl_records(&mut stream, LogWriter::default(), &mut output),
        OutputFormat::LogDebug => process_vsl_records(&mut stream, LogDebugWriter::default(), &mut output),
        OutputFormat::RecordDebug => process_vsl_records(&mut stream, RecordDebugWriter::new(store_config), &mut output),
        OutputFormat::SessionDebug => process_vsl_records(&mut stream, SessionDebugWriter::new(store_config), &mut output),
        OutputFormat::Json => process_vsl_records(&mut stream, SerdeWriter::new(Format::Json, config, store_config), &mut output),
        OutputFormat::JsonPretty => process_vsl_records(&mut stream, SerdeWriter::new(Format::JsonPretty, config, store_config), &mut output),
        OutputFormat::NcsaJson => process_vsl_records(&mut stream, SerdeWriter::new(Format::NcsaJson, config, store_config), &mut output),
    }
}

#[derive(Default)]
struct LogWriter;
impl WriteRecord for LogWriter {
    fn write_record<W>(&mut self, record: VslRecord, output: &mut W) -> Result<(), ProcessingError> where W: Write {
        writeln!(output, "{:#}", record).map_err(From::from)
    }
}

#[derive(Default)]
struct LogDebugWriter;
impl WriteRecord for LogDebugWriter {
    fn write_record<W>(&mut self, record: VslRecord, output: &mut W) -> Result<(), ProcessingError> where W: Write {
        writeln!(output, "{:#?}", record).map_err(From::from)
    }
}

struct RecordDebugWriter {
    state: RecordState,
}

impl RecordDebugWriter {
    fn new(store_config: &StoreConfig) -> RecordDebugWriter {
        RecordDebugWriter {
            state: RecordState::with_config(store_config)
        }
    }
}

impl WriteRecord for RecordDebugWriter {
    fn write_record<W>(&mut self, record: VslRecord, output: &mut W) -> Result<(), ProcessingError> where W: Write {
        if let Some(record) = self.state.apply(&record) {
            writeln!(output, "{:#?}", record).map_err(From::from)
        } else {
            Ok(())
        }
    }
}

struct SessionDebugWriter {
    state: SessionState,
}

impl SessionDebugWriter {
    fn new(store_config: &StoreConfig) -> SessionDebugWriter {
        SessionDebugWriter {
            state: SessionState::with_config(store_config)
        }
    }
}

impl WriteRecord for SessionDebugWriter {
    fn write_record<W>(&mut self, record: VslRecord, output: &mut W) -> Result<(), ProcessingError> where W: Write {
        if let Some(record) = self.state.apply(&record) {
            writeln!(output, "{:#?}", record).map_err(From::from)
        } else {
            Ok(())
        }
    }
}

struct SerdeWriter {
    state: SessionState,
    format: Format,
    config: Config,
}

impl SerdeWriter {
    fn new(format: Format, config: Config, store_config: &StoreConfig) -> SerdeWriter {
        SerdeWriter {
            state: SessionState::with_config(store_config),
            format: format,
            config: config,
        }
    }
}

impl WriteRecord for SerdeWriter {
    fn write_record<W>(&mut self, record: VslRecord, output: &mut W) -> Result<(), ProcessingError> where W: Write {
        if let Some(client) = self.state.apply(&record) {
            log_client_record(&client, &self.format, output, &self.config).map_err(From::from)
        } else {
            Ok(())
        }
    }

    fn log_reports(&self) {
        for client in self.state.unresolved_root_client_access_records() {
            warn!("Root ClientAccessRecord left unresolved:\n{:#?}", client)
        }

        for client in self.state.unresolved_client_access_records() {
            warn!("ClientAccessRecord left unresolved:\n{:#?}", client)
        }

        for backend in self.state.unresolved_backend_access_records() {
            warn!("BackendAccessRecord left unresolved:\n{:#?}", backend)
        }
    }
}

arg_enum! {
    #[derive(Debug)]
    enum OutputFormat {
        Log,
        LogDebug,
        RecordDebug,
        SessionDebug,
        Json,
        JsonPretty,
        NcsaJson
    }
}

fn validate_max_record_slots(value: String) -> Result<(), String> {
    let v = try!(value.parse::<usize>().map_err(|_| format!("max-record-slots expected to be an integer; got: {:?}", value)));

    if !(v > 0) {
        Err(format!("max-record-slots must be greater than zero; got: {}", v))
    } else {
        Ok(())
    }
}

fn validate_max_epoch_diff(value: String) -> Result<(), String> {
    let v = try!(value.parse::<u64>().map_err(|_| format!("max-epoch-diff expected to be an integer; got: {:?}", value)));

    if !(v > 0) {
        Err(format!("max-epoch-diff must be greater than zero; got: {}", v))
    } else {
        Ok(())
    }
}

fn validate_evict_factor(value: String) -> Result<(), String> {
    let v = try!(value.parse::<f32>().map_err(|_| format!("evict-factor expected to be an integer; got: {:?}", value)));

    if !(v > 0.0) {
        Err(format!("evict-factor must be greater than zero; got: {}", v))
    } else {
        Ok(())
    }
}

fn validate_stat_epoch_inverval(value: String) -> Result<(), String> {
    let v = try!(value.parse::<u64>().map_err(|_| format!("stat-epoch-interval expected to be an integer; got: {:?}", value)));

    if !(v > 0) {
        Err(format!("stat-epoch-interval must be greater than zero; got: {}", v))
    } else {
        Ok(())
    }
}

fn main() {
    let arguments = App::new("Varnish Structured Logger")
        .version(crate_version!())
        .author(crate_authors!())
        .about("Reads Varnish VSL (binary) log stream and emits JSON log entries to standard output")
        .arg(Arg::with_name("log-spec")
             .short("d")
             .long("log-sepc")
             .value_name("LOG_LEVEL_SPEC")
             .help("Logging level specification, e.g: info")
             .takes_value(true)
             .default_value("info"))
        .arg(Arg::with_name("output")
             .long("output-format")
             .short("o")
             .help("Format of the output")
             .takes_value(true)
             .possible_values(&OutputFormat::variants())
             .default_value(OutputFormat::variants().last().unwrap()))
        .arg(Arg::with_name("no-log-processing")
             .long("no-log-processing")
             .short("l")
             .help("Do not process VSL log into vars, messages and ACL matches"))
        .arg(Arg::with_name("keep-raw-log")
             .long("keep-raw-log")
             .short("L")
             .help("Include raw log messages"))
        .arg(Arg::with_name("no-header-indexing")
             .long("no-header-indexing")
             .short("i")
             .help("Do not make indices of request and response headers with normalized header names"))
        .arg(Arg::with_name("keep-raw-headers")
             .long("keep-raw-headers")
             .short("I")
             .help("Keep raw header name/value pairs; any indices are moved to top level"))
        .arg(Arg::with_name("stat-epoch-interval")
             .long("stat-epoch-interval")
             .short("s")
             .help("Log store stats every epoch interval (~seconds) on record insert")
             .takes_value(true)
             .validator(validate_stat_epoch_inverval)
             .default_value("600"))
        .arg(Arg::with_name("stream-buffer-size")
             .long("stream-buffer-size")
             .display_order(2000)
             .help("Size of stream buffer in bytes - must be bigger than biggest VSL record")
             .default_value("262144"))
        .arg(Arg::with_name("max-record-slots")
             .long("max-record-slots")
             .display_order(2000)
             .help("Maximum number of records being built or correlated at the same time")
             .validator(validate_max_record_slots)
             .default_value("4000"))
        .arg(Arg::with_name("max-epoch-diff")
             .long("max-epoch-diff")
             .display_order(2000)
             .help("Consider record being built or correlated as stale after that many epoch (~seconds) elapsed")
             .validator(validate_max_epoch_diff)
             .default_value("43200"))
        .arg(Arg::with_name("evict-factor")
             .long("evict-factor")
             .help("Remove that many records at a time when doing expiration or nucking (ratio to max-record-slots)")
             .display_order(2000)
             .validator(validate_evict_factor)
             .default_value("0.01"))
        .arg(Arg::with_name("vsl-file")
             .value_name("VSL_FILE")
             .help("VSL file to process (read from standard input if not specified)"))
        .get_matches();

    program::init(arguments.value_of("log-spec"));

    let output_format = value_t!(arguments, "output", OutputFormat).unwrap_or_else(|e| e.exit());
    let stream_buf_size = value_t!(arguments, "stream-buffer-size", usize).unwrap_or_else(|e| e.exit());

    let output = std::io::stdout();

    let config = Config {
        no_log_processing: arguments.is_present("no-log-processing"),
        keep_raw_log: arguments.is_present("keep-raw-log"),
        no_header_indexing: arguments.is_present("no-header-indexing"),
        keep_raw_headers: arguments.is_present("keep-raw-headers"),
    };

    let stat_epoch_interval = if arguments.is_present("stat-epoch-interval") {
        Some(value_t!(arguments, "stat-epoch-interval", u64).unwrap_or_else(|e| e.exit()))
    } else {
        None
    };

    let _epoch_timer = spawn_epoch_timer();

    let store_config = StoreConfig::new(
        value_t!(arguments, "max-record-slots", usize).unwrap_or_else(|e| e.exit()),
        value_t!(arguments, "max-epoch-diff", u64).unwrap_or_else(|e| e.exit()),
        value_t!(arguments, "evict-factor", f32).unwrap_or_else(|e| e.exit()),
        stat_epoch_interval,
        Some(global_epoch)
    ).unwrap();

    let result = if let Some(path) = arguments.value_of("vsl-file") {
        let file = File::open(path);
        match file {
            Ok(file) => process_vsl_stream(file, output, stream_buf_size, output_format, config, &store_config),
            Err(err) => program::exit_with_error(&format!("Failed to open VSL file: {}: {}", path, err), 1),
        }
    } else {
        let stdin = stdin();
        let stdin = stdin.lock();
        process_vsl_stream(stdin, output, stream_buf_size, output_format, config, &store_config)
    };

    if let Err(err) = result {
        if err.is_brokend_pipe() {
            info!("Broken pipe")
        } else {
            error!("{}", err);
            program::exit_with_error(err.description(), err.to_exit_code())
        }
    }

    info!("Done");
}
